test/normalCommand.test.ts

```diff
@@ -0,0 +1,74 @@
+import { newTest } from './testSimplifier';
+import { cleanUpWorkspace, setupWorkspace } from './testUtils';
+
+suite('Execute normal command', () => {
+  setup(async () => {
+    await setupWorkspace();
+  });
+
+  teardown(cleanUpWorkspace);
+
+  newTest({
+    title: 'One liner',
+    start: ['foo =| bar = 1'],
+    keysPressed: ':normal f=i!=\n',
+    end: ['foo = bar !|== 1'],
+  });
+
+  newTest({
+    title: 'One liner with selection',
+    start: ['foo =| bar = 1'],
+    keysPressed: 'V:normal f=i!=\n',
+    end: ['foo !|== bar = 1'],
+  });
+
+  newTest({
+    title: 'Multiple liner with selection',
+    start: ['foo =| bar = 1', 'foo = bar = 2'],
+    keysPressed: 'Vj:normal f=i!=\n',
+    end: ['foo !== bar = 1', 'foo !|== bar = 2'],
+  });
+
+  newTest({
+    title: 'Multiple liner with line range',
+    start: ['foo =| bar = 1', 'foo = bar = 2', 'foo = bar = 3'],
+    keysPressed: ':2,3normal f=i!=\n',
+    end: ['foo = bar = 1', 'foo !== bar = 2', 'foo !|== bar = 3'],
+  });
+
+  newTest({
+    title: 'One liner with dot',
+    start: ['foo =| bar = 1', 'foo = bar = 2'],
+    keysPressed: 'f=i!=<Esc>j^f=:normal .\n',
+    end: ['foo = bar !== 1', 'foo !|== bar = 2'],
+  });
+
+  // TODO: remove comment out after fixing dot command bug
+  // newTest({
+  //   title: 'One liner with multiple dot',
+  //   start: ['foo =| bar = 1', 'foo = bar = 2'],
+  //   keysPressed: 'f=i!=<Esc>j^f=:normal 2.\n',
+  //   end: ['foo = bar !== 1', 'foo !=!|== bar = 2'],
+  // });
+
+  newTest({
+    title: 'One liner with macro',
+    start: ['|1. one, 2. two, 3. three, 4. four'],
+    keysPressed: 'qaf.r)q:normal @a\n',
+    end: ['1) one, 2|) two, 3. three, 4. four'],
+  });
+
+  newTest({
+    title: 'One liner with multiple macro',
+    start: ['|1. one, 2. two, 3. three, 4. four'],
+    keysPressed: 'qaf.r)q:normal 3@a\n',
+    end: ['1) one, 2) two, 3) three, 4|) four'],
+  });
+
+  newTest({
+    title: 'Multiple liner with multiple macro',
+    start: ['|0. zero', '1. one, 2. two, 3. three, 4. four', '5. five, 6. six, 7. seven, 8. eight'],
+    keysPressed: 'qaf.r)qjVj:normal 4@a\n',
+    end: ['0) zero', '1) one, 2) two, 3) three, 4) four', '5) five, 6) six, 7) seven, 8|) eight'],
+  });
```

Comment:
Add a test that runs multiple commands via `:normal` and then runs `u` - the whole thing should be undone.

Reply:
Fix it.
[c574554](https://github.com/VSCodeVim/Vim/pull/8896/commits/c57455412517f0941c2816ebceb0cde5f894a6ae)

When adding this test, I found that the cursor position is not back because `historyTracker.finishCurrentStep()` is not called after finishing executing normal command.
To fix it, I introduced NormalCommandState to detect end of normal command finish.

I'm not sure this is the best solution.
Please give me some feedbacks!!

Reply:
Yeah, I think your approach makes sense